src/vma/lwip/tcp_out.c

```diff
@@ -852,120 +851,128 @@ tcp_send_empty_ack(struct tcp_pcb *pcb)
 #define TCP_HDR  (20)
 
 void
-tcp_split_segment(struct tcp_pcb *pcb, struct tcp_seg *seg, u32_t wnd){
-
-	struct pbuf *p = NULL;
-	struct tcp_seg *newseg = NULL;
-	u32_t lentosend = (wnd - (seg->seqno - pcb->lastack));
-	u16_t oversize = 0;
-	u8_t  optlen = 0;
-	u8_t  apiflags = 0;
-	u8_t  optflags = 0;
-
-	if (seg->p->ref>1) {
-		return;
-	}
+tcp_split_segment(struct tcp_pcb *pcb, struct tcp_seg *seg, u32_t wnd)
+{
 
-	#if LWIP_TCP_TIMESTAMPS
-	if ((pcb->flags & TF_TIMESTAMP)) {
-		optflags = TF_SEG_OPTS_TS;
-	}
-	#endif /* LWIP_TCP_TIMESTAMPS */
+  struct pbuf *p = NULL;
+  struct tcp_seg *newseg = NULL;
+  u32_t lentosend = (wnd - (seg->seqno - pcb->lastack));
+  u16_t oversize = 0;
+  u8_t  optlen = 0;
+  u8_t  apiflags = 0;
+  u8_t  optflags = 0;
 
-	optlen = LWIP_TCP_OPT_LENGTH( optflags );
+  if ((NULL == seg->p) || (seg->p->ref>1)) {
+	  return;
+  }
 
-	if(seg->p->len > lentosend) {//first buffer is too big, split it
+#if LWIP_TCP_TIMESTAMPS
+  if ((pcb->flags & TF_TIMESTAMP)) {
+    optflags = TF_SEG_OPTS_TS;
+  }
+#endif /* LWIP_TCP_TIMESTAMPS */
 
-		u32_t lentoqueue = seg->p->len - TCP_HDR - lentosend;
+  optlen = LWIP_TCP_OPT_LENGTH( optflags );
 
-		if (NULL == (p = tcp_pbuf_prealloc(lentoqueue + optlen, lentoqueue + optlen, &oversize, pcb, apiflags, 0))) {
-			LWIP_DEBUGF(TCP_OUTPUT_DEBUG | 2, ("tcp_output : could not allocate memory for pbuf copy size %"U16_F"\n", (lentoqueue + optlen)));
-			return;
-		}
+  if (seg->p->len > lentosend) {//first buffer is too big, split it
 
-		/* Copy the data from the original buffer */
-		TCP_DATA_COPY2((char *)p->payload + optlen, (u8_t *)seg->dataptr + lentosend, lentoqueue , &chksum, &chksum_swapped);
+    u32_t lentoqueue = seg->p->len - TCP_HDR - lentosend;
```

Comment:
What guarantees that lentoqueue is non-negative? If it is negative - as unsigned we'll get a big number